### PR TITLE
fix missing input checks in jax.nn.functions

### PR DESCRIPTION
**Pull request by DanisNone**

### CODE REVIEW

**Comment by jakevdp** [comment]
Thanks! We do want this kind of check in our public APIs, though techincally this is an API breakage (previously valid code will begin resulting in an error). For that reason we have the `emit_warning` argument in `check_arraylike`, which can be set to `True` for any new checks. Then after a suitable deprecation period we can change the warning to an error.

**Comment by DanisNone** [comment]
> Thanks! We do want this kind of check in our public APIs, though techincally this is an API breakage (previously valid code will begin resulting in an error). For that reason we have the `emit_warning` argument in `check_arraylike`, which can be set to `True` for any new checks. Then after a suitable deprecation period we can change the warning to an error.

Do I need to set `emit_warning=True`?
If yes, it doesn’t always make sense. For example, in `relu`, there’s a call to `maximum`, which performs a similar check.  After adding `check_arraylike`, the error message will mention `relu` rather than `maximum`.

**Comment by DanisNone** [comment]
I added `emit_warning=True` in `squareplus`, `leaky_relu`, and `elu` since they use `jnp.asarray`. Also added it in `softmax` due to its complex logic, and in `sigmoid` because `lax.logistic` lacks internal checks.

File: jax/_src/nn/functions.py [hunk 1]
```diff
@@ -121,8 +122,7 @@ def squareplus(x: ArrayLike, b: ArrayLike = 4) -> Array:
     x : input array
     b : smoothness parameter
   """
-  numpy_util.check_arraylike("squareplus", x)
-  numpy_util.check_arraylike("squareplus", b)
+  numpy_util.check_arraylike("squareplus", x, b, emit_warning=True)
```
Comment (jakevdp) [inline jax/_src/nn/functions.py]: Since this function checked the validity of inputs previously, we should not set `emit_warning=True`.

File: jax/_src/nn/functions.py [hunk 2]
```diff
@@ -203,6 +204,7 @@ def sigmoid(x: ArrayLike) -> Array:
     :func:`log_sigmoid`
 
   """
+  numpy_util.check_arraylike("sigmoid", x, emit_warning=True)
```
Comment (jakevdp) [inline jax/_src/nn/functions.py]: I think removing `emit_warning=True` is safe here, becuase `lax.logistic` requires array-like input.

File: jax/_src/nn/functions.py [hunk 3]
```diff
@@ -330,7 +333,7 @@ def elu(x: ArrayLike, alpha: ArrayLike = 1.0) -> Array:
   See also:
     :func:`selu`
   """
-  numpy_util.check_arraylike("elu", x)
+  numpy_util.check_arraylike("elu", x, alpha, emit_warning=True)
```
Comment (jakevdp) [inline jax/_src/nn/functions.py:337]: Here I would do
```
numpy_util.check_arraylike("elu", x)
numpy_util.check_arraylike("elu", x, alpha, emit_warning=True)
```
because we want to continue to error instead of warn if `x` is not an array

File: jax/_src/nn/functions.py [hunk 4]
```diff
@@ -360,7 +363,7 @@ def leaky_relu(x: ArrayLike, negative_slope: ArrayLike = 1e-2) -> Array:
   See also:
     :func:`relu`
   """
-  numpy_util.check_arraylike("leaky_relu", x)
+  numpy_util.check_arraylike("leaky_relu", x, negative_slope, emit_warning=True)
```
Comment (jakevdp) [inline jax/_src/nn/functions.py:368]: Same comment here as in `elu`.

File: jax/_src/nn/functions.py [hunk 5]
```diff
@@ -588,6 +593,7 @@ def softmax(x: ArrayLike,
   See also:
     :func:`log_softmax`
   """
+  numpy_util.check_arraylike("softmax", x, emit_warning=True)
```
Comment (jakevdp) [inline jax/_src/nn/functions.py]: We can remove `emit_warning` here, becuase `x` is currently validated when passed to `max`.
